### PR TITLE
Preset courant functions for atmos Model

### PR DESCRIPTION
**Pull request by slifer50**
The PR defines a module for some preset courant functions that make use of #618. These currently enable non-diffusive and diffusive courant number calculations in the vertical or horizontal directions.

### CODE REVIEW

File: src/Courant/Courant_presets.jl [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  k̂ = aux.orientation.∇Φ / norm(aux.orientation.∇Φ)
+  if direction isa VerticalDirection
+    u = dot(state.ρu/state.ρ, k̂) * k̂
+          return Δt * (norm(u) + soundspeed(m.moisture, m.orientation, state,aux)) / Δx
+  elseif direction isa HorizontalDirection
+    u = cross(k̂, cross(state.ρu/state.ρ,k̂))      
+    return Δt * (norm(u) + soundspeed(m.moisture, m.orientation, state,
+                                            aux)) / Δx
+  end
+end
```
Comment (smarras79) [inline src/Courant/Courant_presets.jl]: This is very clean and looks good. 

Can we have a third function just to calculate the advective (i.e. without soundspeed) courant?

File: src/Courant/Courant_presets.jl [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+function Diffusive_CFL(m::AtmosModel, state::Vars, aux::Vars,
+                               diffusive::Vars, Δx, Δt, direction=VerticalDirection())
+  ν, τ, sdiff = turbulence_tensors(m.turbulence, state, diffusive, aux, 0)
```
Comment (smarras79) [inline src/Courant/Courant_presets.jl]: Can't we pass the value of the eddy viscosity which is already computed elsewhere instead of re-computing it?

File: src/Courant/Courant_presets.jl (hunk 1 above)
Comment (slifer50) [reply src/Courant/Courant_presets.jl]: Yes I'll add that in shortly

File: src/Courant/Courant_presets.jl (hunk 2 above)
Comment (slifer50) [reply src/Courant/Courant_presets.jl]: I'm looking into seeing if I can do this with update-diffusive.

Comment (jkozdon) [reply src/Courant/Courant_presets.jl]: Note that there is no guarantee that the current `diffusive` and `aux` arrays is consistent with the `state` outside of the dg model [1], so for safety reason you may want to recompute these updates before calculating the CFL.

> Can't we pass the value of the eddy viscosity which is already computed elsewhere instead of re-computing it?

It is never stored anymore, so you must recompute it [2].

[1]: https://github.com/climate-machine/CLIMA/blob/7977f64a67cb745f3f23a6c90c0c4ad12328c915/src/DGmethods/DGmodel.jl#L20-L175
[2]: https://github.com/climate-machine/CLIMA/blob/c117b4c7eb8b33b3133fcd5b4886269d1356bc08/src/Atmos/Model/AtmosModel.jl#L175-L182

File: src/Atmos/Model/turbulence.jl [hunk 3]
```diff
@@ -104,11 +104,14 @@ end
 
 function turbulence_tensors(m::ConstantViscosityWithDivergence,
     state::Vars, diffusive::Vars, aux::Vars, t::Real)
-
+  FT = eltype(state)
+  k̂ = aux.orientation.∇Φ / norm(aux.orientation.∇Φ)
```
Comment (charleskawczynski) [inline src/Atmos/Model/turbulence.jl]: Should this be
```
k̂ = vertical_unit_vector(orientation::Orientation, aux::Vars)
```
from `orientation.jl`? In which case, you'll need to pass in the `AtmosModel`.

If so, please repeat this elsewhere, where needed.

File: src/Courant/Courant_presets.jl [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+using ..Mesh.Grids
+using ..Mesh.Grids: VerticalDirection, HorizontalDirection, EveryDirection
+using Logging
+using Printf
+using LinearAlgebra
+using CLIMA.Atmos
+
+
+
+function Advective_CFL(m::AtmosModel, state::Vars, aux::Vars,
```
Comment (jkozdon) [inline src/Courant/Courant_presets.jl]: Note that both of these definitions of Advection CLF and Diffusive CFL are atmos specific functions. Might be better to have them live in the atmos model and create similar functions for the ocean?

Style-wise `Advective_CFL` and `Diffusive_CFL` should be lower case since they are not types.

Comment (slifer50) [reply src/Courant/Courant_presets.jl]: @jkozdon I see what you mean, I'll make the changes as requested

**Comment by slifer50** [comment]
> @slifer50 Thanks for taking this on.
> 
> There seems to be a bunch of changes unrelated to the PR. probably good to clean that up before merging, and might be good to extract your functionality and add it to the exist #618 PR.

@jkozdon Could you point out the unrelated changes you mentioned? some small changes were made in Atmos Model and turbulence to be able to split the viscosity into horizontal and vertical components as well as test that it was being done correctly.

File: src/Atmos/Model/turbulence.jl (hunk 3 above)
Comment (slifer50) [reply src/Atmos/Model/turbulence.jl]: Does this work as intended with Spherical orientation?

File: experiments/LES/dry_rayleigh_benard.jl [hunk 5]
```diff
@@ -0,0 +1,133 @@
+using Distributions
```
Comment (jkozdon) [inline experiments/LES/dry_rayleigh_benard.jl]: Are these changes needed?

File: experiments/LES/dycoms.jl [hunk 6]
```diff
@@ -63,16 +63,16 @@ function init_dycoms!(state, aux, (x,y,z), t)
...
+    #r1 = FT(rand(Uniform(-0.002, 0.002)))
+    #r2 = FT(rand(Uniform(-0.00001, 0.00001)))
+    #r3 = FT(rand(Uniform(-0.001, 0.001)))
+    #r4 = FT(rand(Uniform(-0.001, 0.001)))
+    #if z <= 400.0
+     #   θ_liq += r1 * θ_liq
+      #  q_tot += r2 * q_tot
+       # u     += r3 * u
+       # v     += r4 * v
+    #end
```
Comment (jkozdon) [inline experiments/LES/dycoms.jl]: Are these changes needed?

File: experiments/LES/dycoms.jl [hunk 7]
```diff
@@ -140,7 +140,7 @@ function config_dycoms(FT, N, resolution, xmax, ymax, zmax)
                                      solver_type=CLIMA.ExplicitSolverType(LSRK144NiegemannDiehlBusch),
                                      ref_state=ref_state,
                                      C_smag=C_smag,
-                                     moisture=EquilMoist(5),
+                                     moisture=EquilMoist(),
```
Comment (jkozdon) [inline experiments/LES/dycoms.jl]: Are these changes needed?

File: src/Atmos/Model/turbulence.jl (hunk 3 above)
Comment (jkozdon) [reply src/Atmos/Model/turbulence.jl]: Can we move this whole computation to the CFL computation? The reason I suggest this is that we changing the interface and computation for just the CFL (which isn't called at every time step but only occasionally).

File: src/Atmos/Model/turbulence.jl [hunk 8]
```diff
@@ -89,7 +89,7 @@ $(DocStringExtensions.FIELDS)
 """
 struct ConstantViscosityWithDivergence{FT} <: TurbulenceClosure
   "Dynamic Viscosity [kg/m/s]"
-  ρν::FT
+  ρν::SVector{3,FT}
```
Comment (jkozdon) [inline src/Atmos/Model/turbulence.jl]: Why is this needed?

File: src/Atmos/Model/AtmosModel.jl [hunk 9]
```diff
@@ -174,7 +174,7 @@ end
 
 @inline function flux_diffusive!(atmos::AtmosModel, flux::Grad, state::Vars,
                                  diffusive::Vars, aux::Vars, t::Real)
-  ν, τ = turbulence_tensors(atmos.turbulence, state, diffusive, aux, t)
+  ν, τ, sdiff = turbulence_tensors(atmos.turbulence, state, diffusive, aux, t)
```
Comment (jkozdon) [inline src/Atmos/Model/AtmosModel.jl]: See comment below

File: experiments/LES/dycoms.jl (hunk 7 above)
Comment (charleskawczynski) [reply experiments/LES/dycoms.jl]: This will break dycoms, please revert this.

File: src/Atmos/Model/turbulence.jl (hunk 3 above)
Comment (jkozdon) [reply src/Atmos/Model/turbulence.jl]: The reason this is a concern really is that we might kill our register usage on the GPU doing stuff that isn't required.

File: experiments/LES/dry_rayleigh_benard.jl (hunk 5 above)
Comment (slifer50) [reply experiments/LES/dry_rayleigh_benard.jl]: No I will remove this, thanks.

File: experiments/LES/dycoms.jl (hunk 6 above)
Comment (slifer50) [reply experiments/LES/dycoms.jl]: No, I'll remove this.

File: experiments/LES/dycoms.jl (hunk 7 above)
Comment (slifer50) [reply experiments/LES/dycoms.jl]: No I'll clean that up.

File: src/Atmos/Model/turbulence.jl (hunk 8 above)
Comment (slifer50) [reply src/Atmos/Model/turbulence.jl]: This was done for testing purposes, so we could set different constant viscosities in different directions.

File: src/Atmos/Model/AtmosModel.jl (hunk 9 above)
Comment (slifer50) [reply src/Atmos/Model/AtmosModel.jl]: I'm currently looking for an effective way of not having this. Since we need a split viscosity for the courant numbers and non split viscosity for the general calculations. "sdiff" is needed  exclusively for the courant calculations.

Comment (slifer50) [reply src/Atmos/Model/AtmosModel.jl]: Since it doesn't look like I can avoid having to do a recomputation I think I will write separate function in turbulence to get the split viscosity.

File: src/Atmos/Model/turbulence.jl [hunk 10]
```diff
@@ -241,14 +250,15 @@ end
 function diffusive!(::Vreman, orientation::Orientation,
                     diffusive::Vars, ∇transform::Grad, state::Vars, aux::Vars, t::Real)
   diffusive.turbulence.∇u = ∇transform.u
+  diffusive.turbulence.S = symmetrize(diffusive.turbulence.∇u)
```
Comment (jkozdon) [inline src/Atmos/Model/turbulence.jl]: This is storing the same thing twice, which will increase our GPU memory usage. Possibly we should just be storing the symmeterized version?

Comment (slifer50) [reply src/Atmos/Model/turbulence.jl]: The Vreman subgridscale model seems to need both, however it computes the symmeterized version locally. However I currently need S to be accessible to do the Courant calculation. This would be simple if all the subgridscale models either typically stored only of the two tensors but this seems to vary depending on the choice of turbulence closure. It might be best in this case to possibly only store the non symmeterized version for all closures and have them compute the symmetrized version locally so that it is never saved before this is we must be certain it is not needed elsewhere.

File: src/Atmos/Model/turbulence.jl (hunk 3 above)
Comment (slifer50) [reply src/Atmos/Model/turbulence.jl]: @jkozdon noted, I will attempt to move this into the CFL computation itself as you suggest.

File: src/Atmos/Model/turbulence.jl (hunk 10 above)
Comment (jkozdon) [reply src/Atmos/Model/turbulence.jl]: Recomputing simple algebraic quantities is almost alway preferential in general (loading data is the most expensive thing we do, so loading the same data multiple times in the same functions doesn't make sense).

When we start needing to do iterations to computing values or using special functions the picture changes a bit.

File: src/Atmos/Model/turbulence.jl (hunk 8 above)
Comment (slifer50) [reply src/Atmos/Model/turbulence.jl]: Since I've already done tests with this, I can remove this and revert the changes for the constant viscosity closure, would that be alright @jkozdon ?

**Comment by jkozdon** [comment]
> might be good to extract your functionality and add it to the exist #618 PR

Talked to @lcw and he is fine with you pushing commits to his branch.

The other advantage of this is that you would preserve his commit history, which means if code breaks that he wrote we will know who to talk to (right now since you took his code but not his commits it seems that you wrote all the code in the git history).

**Comment by jkozdon** [comment]
> > might be good to extract your functionality and add it to the exist #618 PR
> 
> Talked to @lcw and he is fine with you pushing commits to his branch.

You can also get everything working here then we can create a diff commit that we push to his branch before merging the PR

File: src/Atmos/Model/turbulence.jl (hunk 8 above)
Comment (jkozdon) [reply src/Atmos/Model/turbulence.jl]: I would suggest reverting it then if its not needed in the final PR

File: src/Atmos/Model/turbulence.jl (hunk 10 above)
Comment (slifer50) [reply src/Atmos/Model/turbulence.jl]: It looks S isn't used outside of turbulence.jl I'm going to modify it so that only the non-symmeterized version is saved and S can be computed for each closure as needed.

**Comment by slifer50** [comment]
> > > might be good to extract your functionality and add it to the exist #618 PR
> > 
> > 
> > Talked to @lcw and he is fine with you pushing commits to his branch.
> 
> You can also get everything working here then we can create a diff commit that we push to his branch before merging the PR

Sounds good, we can do that once I've addressed everything here.

File: src/Courant/Courant_presets.jl (hunk 4 above)
Comment (tapios) [reply src/Courant/Courant_presets.jl]: Better yet: make the functions universal here, i.e., calculate the relevant soundspeed and viscosity/diffusivity in the atmosphere/ocean models, then call the functions here with those model-specific quantities. 

Please also change `_CFL` to `Courant_number` for consistency with the rest. 

BTW: we want to share the turbulence module between atmosphere and ocean in the end. The functions we need there are the same for atmosphere and ocean; just parameters are different.

Comment (slifer50) [reply src/Courant/Courant_presets.jl]: @tapios @jkozdon I could probably change the functions to use any BalanceLaw instead of AtmosModel allowing them to be used by any model, would this be alright?

Comment (tapios) [reply src/Courant/Courant_presets.jl]: Yes, that sounds good to me.

**Comment by slifer50** [comment]
@jkozdon @tapios @lcw @charleskawczynski @smarras79 I believe I've adressed everything that's been mentioned so far in the latest 2 commits, please let me know if I've missed something.

File: src/Atmos/Model/turbulence.jl [hunk 11]
```diff
@@ -321,13 +322,17 @@ end
 
 function turbulence_tensors(m::AnisoMinDiss, state::Vars, diffusive::Vars, aux::Vars, t::Real)
   FT = eltype(state)
+end
+function scaled_momentum_flux_tensor(m::AnisoMinDiss, ρν, S)
```
Comment (skandalaCLIMA) [inline src/Atmos/Model/turbulence.jl]: @slifer50 : Hi Yassine, I am working on merging the CFL pull requests #618 & #667. Can I request you to provide a clarification on this revision. 

With this revision, the function turbulence_tensors, just returns FT = eltype(state). This seems to be called in the "courant_presets" as ν, τ = turbulence_tensors(m.turbulence, state, diffusive, aux, 0).  

- please advise if the diffusive_courant function has been approved from the physics/numerics point of view. 
- it may be a good idea to make any revisions to the Turbulence model in a different pull request.

Thanks.
cc: @simonbyrne, @kpamnany @akshaysridhar @bischtob

**Comment by simonbyrne** [comment]
Merged as part of #712.
